### PR TITLE
[AutoSP] (Sequence Parallelism) support for Multimodal Models (ViT + LLM)

### PR DESCRIPTION
**Pull request by nathon-lee**
## Description

Hello DeepSpeed Team! 👋 

This PR directly addresses the **"Multimodal model support"** goal outlined in the **DeepSpeed Roadmap Q2 2026 (#7861)**. 

It introduces **AutoSP (Sequence Parallelism) support for Multimodal Models (ViT + LLM)** out of the box. As noted in the roadmap, multimodal models handle significantly longer sequence lengths, making SP critical. This PR automates the injection of DeepSpeed Ulysses-based sequence parallelism into multimodal architectures, removing the need for manual and error-prone engineering efforts.

This is a consolidated PR of several incremental features developed and thoroughly tested in my fork.

### 🎯 Related Issue
- Addresses the Multimodal model support item in **#7861 (DeepSpeed Roadmap Q2 2026)**.
- Builds upon the AutoSP foundation introduced in #7860.

### 🌟 Key Features & Contributions

1. **AutoSP Scaffolding & Detector (`auto_wrap_model_for_sp`)**:
   - Introduced a scanning utility to automatically detect ViT encoders and LLM decoders within a multimodal model.
   - Automatically wraps LLM decoder attention layers with DeepSpeed's existing `DistributedAttention`.

2. **ViT Sequence Parallelism (`UlyssesSPViTAttention`)**:
   - Implemented a Ulysses-style `Gather-Compute-Scatter` sequence parallel wrapper tailored for non-causal ViT attention layers.
   - Significantly reduces the memory footprint of ViT Feed-Forward Networks (FFN) and LayerNorms across the sequence dimension.

3. **Cross-Modal Fusion Adapters (Phase 2)**:
   - Handled the complex sequence scatter/gather at the vision-language boundary to ensure the LLM decoder receives uniformly sharded fused sequences.
   - Supported architectures include:
     - **LLaVA** (`LlavaFusionAdapter`): Visual token splice replacing image placeholders.
     - **InternVL** (`InternVLFusionAdapter`): `IMG_CONTEXT` token splice.
     - **Qwen2-VL** (`Qwen2VLFusionAdapter`): Vision_start/end bounded splice.

### 🧪 Testing & Validation

To ensure this PR does not break any existing functionality and is numerically sound, comprehensive tests have been added:

- **Numerical Equivalence Tests**: Added multi-GPU tests (`tests/unit/sequence_parallelism/test_autosp_equivalence.py`) verifying that the SP-wrapped path across N ranks produces the **exact same numerical results** as the equivalent single-device (non-SP) computation.
- **Integration Tests**: End-to-end mock integration tests validating the full pipeline from ViT to fusion adapter.
- **Benchmarks Provided**: Included a multimodal SP benchmark script (`benchmarks/autosp/bench_multimodal_sp.py`) to easily verify throughput scaling and peak GPU memory reduction.

*(All tests pass cleanly on 2 GPUs with `NCCL_P2P_DISABLE=1`)*

### 🚧 Known Limitations & Future Work

To be fully transparent, there are a few limitations in the current design that I plan to improve in follow-up iterations (or would love guidance on from the team):

1. **Manual Wrapping for Fusion Layers**: While ViT and LLM attentions are wrapped automatically, the vision projection layer currently requires manual wrapping with `ModalityFusionSPAdapter` due to varying HF model implementations. Fully automating Phase 2 is a logical next step.
2. **ViT SP Trade-off**: The current `UlyssesSPViTAttention` uses a Gather-Compute-Scatter approach. While it successfully reduces FFN memory by $1/N$, it still computes the full attention matrix on every rank. A true All-to-All sequence-to-head transposition for Opaque ViT layers is something I am actively exploring.
3. **Padding Attention Mask**: When `fused_len % world_size != 0`, zero-padding is applied. Currently, the global `attention_mask` is not automatically intercepted and patched, which might require user attention during inference.

---

I would deeply appreciate any feedback or suggestions from the maintainers! I am more than happy to make any required adjustments, refactorings, or add further test cases to get this perfectly aligned with the Q2 roadmap and DeepSpeed's standards. 

Thank you for your time reviewing this! 🚀

### CODE REVIEW

File: deepspeed/sequence/autosp_fusion.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+# DeepSpeed Team
+"""
+ModalityFusionSPAdapter — Phase 2
```
Comment (delock) [inline deepspeed/sequence/autosp_fusion.py:6]: If these adapters intend to be applied in model script, should update documentation to reflect the usage.

File: deepspeed/sequence/autosp_vit.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+Memory benefit: activations *outside* the attention block (e.g. feed-forward
+layers, layer norms) are stored only locally, reducing per-rank memory
+proportional to ``world_size``.
+
+The ``cls`` token (if present) is replicated on every rank and is not split
+across the sequence dimension.  Each rank appends its local patches to the
+same ``cls`` token before calling the wrapped attention.
+
+Padding: when ``num_patches % world_size != 0``, we pad patches with zeros
```
Comment (delock) [inline deepspeed/sequence/autosp_vit.py]: Is there a test covering `num_patches % world_size !=0`?

**Comment by delock** [comment]
Hi @nathon-lee thanks for your contribution!  I have left my review comments, thanks!

**Comment by nathon-lee** [comment]
Hi @delock, thank you for the thorough review!

1. **Documentation**: Absolutely agree — I'll add a concise usage example to 
   `ModalityFusionSPAdapter`'s docstring (and a brief note in the docs) showing 
   how to apply the fusion adapters in a model training script. I'll include 
   this in the current PR.

2. **`num_patches % world_size != 0` test**: Great catch! This edge case is 
   currently untested. I'll add a dedicated test in `test_autosp_equivalence.py` 
   to cover the non-divisible path and verify the padding/strip logic in 
   `UlyssesSPViTAttention.forward()` is correct (the module docstring mentions 
   it, but the implementation needs to be confirmed against the test).

I'll push the updates shortly — thanks again for your time!

File: deepspeed/sequence/autosp_fusion.py (hunk 1 above)
Comment (nathon-lee) [reply deepspeed/sequence/autosp_fusion.py:6]: Usage documentation has been added (lines 24-63 in autosp_fusion.py)

File: deepspeed/sequence/autosp_vit.py (hunk 2 above)
Comment (nathon-lee) [reply deepspeed/sequence/autosp_vit.py]: The test coverage for `test_noneven_patches` has been added, with `num_patches % world_size != 0`

File: deepspeed/sequence/autosp_vit.py [hunk 3]
```diff
@@ -100,10 +101,27 @@ def forward(self, hidden_states: torch.Tensor, **kwargs):
...
+            local_patches_padded = F.pad(local_patches, (0, 0, 0, pad_len))
+        else:
+            local_patches_padded = local_patches
+
+        gathered = [
+            torch.zeros(bs, max_local_len, hidden_dim, dtype=local_patches.dtype, device=local_patches.device)
+            for _ in range(self.world_size)
+        ]
+        dist.all_gather(gathered, local_patches_padded.contiguous(), group=self.process_group)
+        full_patches = torch.cat(gathered, dim=1)  # [bs, world_size * max_local_len, hidden_dim]
```
Comment (delock) [inline deepspeed/sequence/autosp_vit.py]: Hi @nathon-lee , thanks for the fix of uneven divided patches.  I have a follow up question.  I saw paddings added will appear in `full_patches` thus will also appear in full_input.  During attention computation of `full_input`, softmax might be affected by these padding patches.   Should we 'unpad' the allgather result before compute `full_input`, or mask the padding during attention computation?

Comment (nathon-lee) [reply deepspeed/sequence/autosp_vit.py]: Thanks for catching this, @delock — much appreciated. You're right that the zero-padded tokens in `full_patches` would participate in the softmax and lead to divergence from single-device execution.

Fixed by de-padding before calling attention rather than masking. After all_gather, each shard is trimmed to its true length using per-rank lengths collected via a preceding scalar all_gather

**Review by delock — APPROVED** [review]
